testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -880,29 +905,79 @@ impl RelayerThread {
+            &mut self.chainstate.index_conn(),
+            &canonical_stacks_tip,
+            &canonical_stacks_tip_ch,
+        )
+        .map_err(|e| {
+            error!(
+                "Relayer: Failed to get tenure-start block header for stacks tip {canonical_stacks_tip}: {e:?}"
+            );
+            NakamotoNodeError::ParentNotFound
+        })?
+        .ok_or_else(|| {
+            error!(
+                "Relayer: Failed to find tenure-start block header for stacks tip {canonical_stacks_tip}"
+            );
+            NakamotoNodeError::ParentNotFound
+        })?;
+        let highest_tenure_bhh =
+            BlockHeaderHash(highest_tenure_start_block_header.index_block_hash().0);
+
+        // If we won the last good sortition, then we should extend off of it.
+        let last_good_block_election_snapshot = {
             let ih = self.sortdb.index_handle(&burn_tip.sortition_id);
-            ih.get_last_snapshot_with_sortition(burn_tip.block_height)
+            info!(
+                "Relayer: Getting last snapshot with sortition for {}",
+                burn_tip.block_height
+            );
+
+            let sn = ih
+                .get_last_snapshot_with_sortition(burn_tip.block_height)
                 .map_err(|e| {
                     error!("Relayer: failed to get last snapshot with sortition: {e:?}");
                     NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?;
+            if sn.winning_stacks_block_hash != highest_tenure_bhh {
+                info!(
+                    "Relayer: Sortition winner is not committed to the canonical tip; allowing last miner to extend";
+                    "burn_block_height" => burn_tip.block_height,
+                    "consensus_hash" => %burn_tip.consensus_hash,
+                );
+
+                SortitionDB::get_block_snapshot_consensus(
+                    self.sortdb.conn(),
+                    &canonical_stacks_tip_ch,
+                )
+                .map_err(|e| {
+                    error!("Relayer: failed to get block snapshot for canonical tip: {e:?}");
+                    NakamotoNodeError::SnapshotNotFoundForChainTip
                 })?
+                .ok_or_else(|| {
+                    error!("Relayer: failed to get block snapshot for canonical tip");
+                    NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?
+            } else {
+                sn
+            }
         };
 
-        let won_last_sortition = last_block_election_snapshot.miner_pk_hash == Some(mining_pkh);
-        debug!(
-            "Relayer: Current burn block had no sortition. Checking for tenure continuation.";
+        let won_last_sortition =
+            last_good_block_election_snapshot.miner_pk_hash == Some(mining_pkh);
+        info!(
+            "Relayer: Current burn block had no sortition or a bad sortition. Checking for tenure continuation.";
```

Comment:
Is this true? Looking at line 962, the case where `sn.winning_stacks_block_hash == highest_tenure_bhh` - wouldn't that mean the current burn block has a good sortition? Or is the fact that we're in `continue_tenure` mean that my assertion is wrong? Sorry, it's tricky to follow all the code paths here.

Reply:
Right, we're trying to find the last good sortition, given that the current one is already bad since we arrived in `continue_tenure`. We get the last snapshot with a sortition, but if we determine that is bad, then we go back to the sortition of the tip.